feat: Board Service 개발

kingseungwooya opened the pull request:
# Describe your changes
Board Service 구현 
properties -> yml 수정 
dependency jsoup 추가 

jsoup을 이용한 웹페이지 크롤링 
메뉴이름별 enum type으로 정의해  해당 메뉴 url저장
Spring Schedular를 이용해 1시간마다 Board Table refersh후 새로 크롤링 해옴

## Issue ticket number and link (optional)
#8 #7 #6 
## Checklist

- [x] 🤔 이 프로젝트의 스타일 가이드를 따르나요?
- [x] 🤔 머지하기 전에 스스로 코드에 문제가 없음을 확인했나요?
- [x] 🤔 필요한 주석을 필요한 곳에 넣어주었나요?

## Next Step Todo (optional)
Board 글 내용 크롤링 
Bus api 개발 


## Questions

- 💬 질문 사항이에요!
- 데이터베이스 서버 생성 언제쯤 가능할까요?

kingseungwooya committed:
feat: Board Entity 설계

Board Entity설계와 enumtype으로 크롤링 url 열거

kingseungwooya committed:
rename: yml

properties-> yml
yml 설정

kingseungwooya committed:
feat: Board Service

add: jsoup dependency
,Board service

kingseungwooya committed:
feat: Schedular

Board Service에 Schedular 적용
main에서 시작과 즉시 크롤링 해와야 하기에 CommandLineRunner구현

k-gn commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,19 @@
+spring:
+  datasource:
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: jdbc:mysql://localhost:3306/cnubot?useSSL=false&useUnicode=true&serverTimezone=Asia/Seoul
+    username: root
+    password: 1234
```
실무 프로젝트는 아니지만 Public Repository 에서 이렇게 데이터베이스 정보가 노출되는 것은 좋지 않습니다!

yml 자체를 암호화 하거나, 빌드 시 입력값 (Arguments) 을 읽어오는 방식 등 다양한 민감 정보 관리 방법이 존재하니 나중에 한번 찾아 보시면 좋을 것 같습니다~!

kingseungwooya replied:
감사합니다. 참조하여 반영하겠습니다.!

kingseungwooya committed:
feat: swagger

fix: error BoardService

okinawaa commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+  datasource:
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: jdbc:mysql://localhost:3306/cnubot?useSSL=false&useUnicode=true&serverTimezone=Asia/Seoul
+    username: root
+    password: 1234
```
퍼블릭으로 공개된 레포지토리라서 
보안정보는 숨기는게 좋지 않을까요?

kingseungwooya replied:
local서버의 DB라서 gitignore에 추가하지 않았습니다! 추후 RDS서버를 이용하게 된다면 보안정보를 숨길려고 합니다!

okinawaa replied:
좋습니다~! 
저희는 RDS는 이용하지 않고,
ec2내에서 mysql을 직접 구축해서 사용합니다! 
어제 전달드린 정보 사용하시면 될 것 같습니다~🚀

kingseungwooya replied:
감사합니다!!!!! 말씀드린 정보로 추가적인 Issue에 대해서 Contribute하겠습니다

k-gn commented on src/main/java/com/cnubot/cnubotserver/domain/service/BoardServiceImpl.java:
```diff
@@ -0,0 +1,33 @@
+
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class BoardServiceImpl implements BoardService{
+    @Autowired
+    BoardRepository repository;
+
+    BoardCrawling boardCrawling;
```
@Autowired 을 사용한 필드 주입보단 생성자 주입을 사용하는 것을 권장드립니다. 현업에선 생성자 주입을 많이 사용하고 있는 추세입니다. 생성자 주입으로 순환 참조를 예방한다던가, final 키워드를 통한 불변처리 등 다양한 이점을 가져올 수 있습니다!

추가적으로 @RequiredArgsConstructor 를 사용하신다면

```java
private final BoardRepository repository;
private final BoardCrawling boardCrawling;
```

이렇게 불변성을 지키며 쉽게 주입 받을 수 있습니다!

kingseungwooya replied:
감사합니다!! DI부분에 대해서 다시 한번 알아보겠습니다!

k-gn commented on src/main/java/com/cnubot/cnubotserver/domain/entity/DepthSecond.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * STUDENT RECRUITING, EVENT, JOB&CAREER, FORM&DATA
+ */
+public enum DepthSecond {
+    NEWS("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0701&site_dvs_cd=kr&menu_dvs_cd=0701") //번호, 제목, 작성자, 작성일, 조회수, 파일여부
+    , ACADEMIC_INFO("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0702&site_dvs_cd=kr&menu_dvs_cd=0702") //번호, 제목, 작성자, 작성일, 조회수, 파일여부
+    , EDUCATION_INFO("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0704&site_dvs_cd=kr&menu_dvs_cd=0704")
+    , BUSINESS_GROUP_STARTUP("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0709&site_dvs_cd=kr&menu_dvs_cd=0709")
+    , UNIV_CULTURE_FIELD("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_070801&site_dvs_cd=kr&menu_dvs_cd=070801")
+    , CNU_MARKET("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_070802&site_dvs_cd=kr&menu_dvs_cd=070802")
+    , STUDY_COMPETITION("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_070808&site_dvs_cd=kr&menu_dvs_cd=070808")//
+    , RECRUIT_INVITATION("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0705&site_dvs_cd=kr&menu_dvs_cd=0705")//번호, 제목, 작성자, 모집기한, 작성일, 조회수
+    , JOB_SEARCH("https://plus.cnu.ac.kr/_prog/recruit/?menu_dvs_cd=07080401&site_dvs_cd=kr") //번호, 제목, 작성자, 모집기한, 작성일, 조회수
+    , JOB_OFFER("https://plus.cnu.ac.kr/_prog/recruit/?site_dvs_cd=kr&menu_dvs_cd=07080402&gubun=2")//번호, 제목, 작성자, 모집기한, 작성일, 조회수
+    , CNU_NEWS("https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0703&site_dvs_cd=kr&menu_dvs_cd=0703")
+    , STUDENT_RECRUIT("https://cnuint.cnu.ac.kr/cnuint/notice/recruit.do")
+    , EVENT("https://cnuint.cnu.ac.kr/cnuint/notice/event.do")
+    , JOB_CAREER("https://cnuint.cnu.ac.kr/cnuint/notice/job.do")
+    , FORM_DATA("https://cnuint.cnu.ac.kr/cnuint/notice/form.do");
+
+    private String port;
+    private String url;
+    public String getPort(){
+        return port;
+    }
+    DepthSecond(String port) {
+        this.port=port;
+        this.url="https://plus.cnu.ac.kr/_prog/_board/?code=sub07_0704&site_dvs_cd=kr&menu_dvs_cd=";
+    }
+}
```
이렇게 카테고리들을 enum 으로 관리하는 것 좋은 생각인 것 같습니다~ 👍

k-gn reviewed (commented):
전체적으로 코드의 포맷이나 컨벤션이 제각각이라 가독성이 떨어지는 것 같습니다!
보통 코드 포맷(lint) 같은건 사용하는 툴의 기능을 적극 활용하시는 게 좋습니다!
코드를 가독성 좋게 작성하는 것이 나중에 본인이 직접 해당 코드를 다시 볼때도 훨씬 편하게 볼 수 있기 때문에 유지보수 측면에도 좋습니다!

한번 본인만에 기준을 만들어보시는 걸 추천드립니다~~

---

.idea 나 build 같은 종류의 폴더는 깃허브에 안올라와도 되는 불필요한 폴더들입니다! 또한 안에 민감한 정보를 포함할 수 있기 때문에 이러한 폴더들은 gitignore 을 통해 관리하는 것이 중요합니다!

---

시간날 때 기능 확인 후 추가 리뷰하겠습니다~

kingseungwooya commented:
좋은 리뷰 감사합니다!!.

"전체적으로 코드의 포맷이나 컨벤션이 제각각이라 가독성이 떨어지는 것 같습니다!
보통 코드 포맷(lint) 같은건 사용하는 툴의 기능을 적극 활용하시는 게 좋습니다!"
-> 이 부분에 대해서 제가 잘 이해를 못해서 그런데 혹시 추천하시는 레퍼런스가 있을까요?

k-gn commented:
어떤건 변수명이 camelCase고, 어떤건 snake_case 로 지정할 건지
메소드명은 어떤식으로 선언할 건지
띄어쓰기나 탭을 몇번 할 건지
줄 바꿈은 언제하고 언제 안할 건지 등의 내용이 코드 포맷과 컨벤션 개념입니다!

위의 내용을 본인만의 또는 팀원간의 기준을 정해서 개발하면 사소한 충돌 없이 가독성 좋은 개발을 진행하실 수 있습니다!
코드의 냄새를 잡아 코드의 품질을 높인다고 말하는 과정이며 귀찮긴 하지만 습관을 들이면 좋아서 말씀드려보았습니다!

저는 개발 툴을 인텔리제이를 사용하는데 인텔리제이 툴 자체에서 format 설정을 할 수도 있고 저는 인텔리제이에서 제공하는 CheckStyle 이란 플러그인을 사용하여 팀원들과 함께 기준을 정하고 작업하고 있습니다!

더 나아가서 코드 분석을 통해 버그나 취약점, 코드 냄새 등을 잡아주는 코드 분석툴인 소나큐브(sonarQube) 도 있습니다!

kingseungwooya committed:
refactor: Change Injection

Autowired에서 Constructor Injection으로 불변성을 지킴.

kingseungwooya committed:
feat: Add dependency jasypt

kingseungwooya committed:
feat: Set jasypt

DB 정보 보호를 위한 jasypt 설정

kingseungwooya committed:
test: jayspt

VM option으로 argument값 받아서 JasyptEncrypt PASSWORD 암호화 후 테스트 완료

kingseungwooya committed:
refactor: snakecase to camelcase

변수 명 camel case로 통일

kingseungwooya committed:
refactor: refactoring production code

가독성을 높이기 위한 refactoring

kingseungwooya committed:
test: refactoring test

crawling refactoring한 결과 test

kingseungwooya commented:
말씀주신 리뷰 바탕으로 전체적으로 수정해보았습니다!! 확인 부탁드립니다!!

k-gn commented on src/main/java/com/cnubot/cnubotserver/domain/service/BoardCrawling.java:
```diff
@@ -0,0 +1,133 @@
+        if(menu.ordinal()<7){
+            elements = document.select("td");
+            for (int i = 0; i < elementsSize - 6; i += 6) {
+                board.setMenu(menu);
+                board.setBoardNum(elements.get(i).text()); //번호
+                board.setUrl(elements.get(i + 1).childNode(0).attr("href")); //url
+                board.setName(elements.get(i + 1).text()); //제목
+                board.setWriter(elements.get(i + 2).text()); //적송저
+                board.setDate(elements.get(i + 3).text()); //작성일
+                board.setHits(elements.get(i + 4).text()); //조회수
+                repository.save(board);
```
반복되는 코드들인 것 같은데 하나의 메소드로 관리하기 쉽게 분리해보는건 어떨까요?!

kingseungwooya replied:
사실 이부분에 대해서 생성자 인자로 넣을까 고민했었는데 코드가 너무 길어져서 이해가 힘들까봐  setter를 썼었습니다. 괜찮으시다면 생성자인자로 넣어서 줄이는것 말고 더 좋은 방법이 있을까요?

k-gn replied:
4개의 if 에 중복 사용되는 setter 코드가 있어서 말씀드렸습니다! 

menu.ordinal()<7 조건이든 menu.ordinal()<10 조건이든 동일하게 board 에 크롤링한 데이터를 set 하는 과정을 거쳐야 하는 걸로 보이는데 해당 코드를 리팩토링 한번 해보시면 좋을 것 같습니다!

---

```java
Board brd = Board.builder()
.menu(elements.get(i).text())
.name(elements.get(i + 1).text())
.build();
```

이런식의 Builder 패턴을 사용하여 객체를 생성할 수도 있습니다~

k-gn reviewed (commented):
내용 반영하시드라 고생하셨습니다!  👍

kingseungwooya committed:
chore: change encrpt(local -> server)

도커에 올린 ec2서버의 주소와 계정으로 암호화 완료: vm option 양식
-Djasypt.encryptor.password= PW
PW 는 디비 PW와 동일

kingseungwooya committed:
feat: add Builder annotation

생성자 파라미터 주입을 명확하게 하기 위해 빌더패턴 사용

kingseungwooya committed:
fix: fix testcase

테스트케이스 수정

kingseungwooya committed:
refact: change repeating setter to builder pattern

빌더패턴으로 가독성 높임과 반복되는 setter 제거

kingseungwooya committed:
chore: remove blank line

kingseungwooya committed:
fix: fix wrong encrypt url

kingseungwooya committed:
refact: refactor method input type for getViewDetail

getViewDetail함수에서도 getDoc을 사용하기 위해 String형태의 url로 인자 변경

kingseungwooya committed:
fix: fix Board Entity

요구사항에 맞게 게시글의 내용을 추출한 내용을 저장하기 위한 속성 생성

kingseungwooya committed:
feat: add getViewDetail

각 게시글마다 내용을 추출하는 기능
Builder 생성자 객체에 변경된 entity형식에 맞게 boardDetail 추가

kingseungwooya committed:
fix: BoardCrawling Service

새로운 속성 boardDetail을 구하기 위한 보수작업

kingseungwooya committed:
fix: yml

encrpt된 부분 오류 수정
test 로그 확인을 위해 불필요 로그 제거

kingseungwooya committed:
test: getViewDetail

 getViewDetail test

kingseungwooya committed:
fix: entity size

244바이트 넘어가는 내용에 의해 Lob 어노테이션 추가

kingseungwooya committed:
feat: runner

서버 시작시  자동으로 크롤링 데이터 생성

kingseungwooya committed:
refact: deleteAll to deleteAllInBatch

deleteAll 사용시 한 행 삭제할때마다 쿼리문이 날라가기에 deleteAllinBatch로 refactoring

kingseungwooya commented:
✋✋✋✋✋✋추가 기능 사항 ✋✋✋✋✋✋

 ❗1. Board 내용물이 필요하다 해서 그것에 맞게 재구현 했습니다. 
 ❗2. 또한 무분별한 setter을 이용하지 않고 Builder를 이용해서 생 성자에 인자를 주입했습니다. 
 ❗3. 디비 도커에 올린 ec2서버의 주소와 계정으로 암호화 완료: vm option 양식-Djasypt.encryptor.password= PW
PW 는 디비 PW와 동일

Schedular 정상 동작 확인 요구한 양식대로 저장되는지 확인 완료

k-gn commented:
이렇게 차근차근 이런식으로 코드 리팩토링하면서 질을 높이는 과정이 점점 재밌어지셨으면 좋겠네요!!
중복되는 코드를 하나로 줄이는 과정이 저는 재밌더라구요!!

고생하셨습니다~!

kingseungwooya commented:
제가 이번에 우테코 프리코스를 하면서 이런 과정이 주어지더군요. 그 때 생각났습니다. 짧지만 @k-gn 님으로 인해 내가 무엇을 놓치고 있었는지 알 수 있었습니다.좋은 경험하게 해주셔서 정말 감사합니다. 저도 너무 재밌었고 저를 성장케 할 수 있는 과정이었습니다.
